scheduling/taskmetadata.go

```diff
@@ -0,0 +1,533 @@
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	serializedMetadata, err := metadata.Marshal()
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+	err = m.storage.Create(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRunMetaKey := TaskRunMetadataKey{taskID: taskID, runID: metadata.ID, date: startTime}
+	err = m.storage.Create(taskRunMetaKey.String(), string(serializedMetadata))
+	if err != nil {
+		err.AddDetail("task name", name)
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(metadata.ID))
+		return TaskRunMetadata{}, err
+	}
+
+	return metadata, nil
+}
+
+func (m *TaskMetadataManager) GetRunByID(taskID TaskID, runID TaskRunID) (TaskRunMetadata, fferr.GRPCError) {
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskRunMetadata, err := m.storage.Get(taskRunKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskRunMetadata))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	// Want to move this logic out
+	found, runRecord := runs.ContainsRun(runID)
+	if !found {
+		err := fferr.NewKeyNotFoundError(taskRunKey.String(), fmt.Errorf("run not found"))
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	date := runRecord.DateCreated
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: runRecord.RunID, date: date}
+	rec, err := m.storage.Get(taskRunMetadataKey.String())
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+
+	taskRun := TaskRunMetadata{}
+	err = taskRun.Unmarshal([]byte(rec))
+	if err != nil {
+		err.AddDetail("task_id", string(taskID))
+		err.AddDetail("run_id", string(runID))
+		return TaskRunMetadata{}, err
+	}
+	return taskRun, nil
+}
+
+func (m *TaskMetadataManager) GetRunsByDate(start time.Time, end time.Time) (TaskRunList, fferr.GRPCError) {
+	// the date range is inclusive
+	var runs []TaskRunMetadata
+
+	// iterate through each day in the date range including the end date
+	for date := start; date.Before(end) || date.Equal(end); date = date.AddDate(0, 0, 1) {
+		dayRuns, err := m.getRunsForDay(date, start, end)
+		if err != nil {
+			return []TaskRunMetadata{}, err
+		}
+		runs = append(runs, dayRuns...)
+	}
+
+	return runs, nil
+}
+
+func (m *TaskMetadataManager) getRunsForDay(date time.Time, start time.Time, end time.Time) ([]TaskRunMetadata, fferr.GRPCError) {
```

Comment:
What is this for?

- How is a day defined? utc? what about timezones?

Reply:
the key contains a date so this would get the runs for a particular day